### PR TITLE
Refactor adot-operator-images-mirror to use docker go sdk.

### PR DESCRIPTION
**Pull request by jefchien**
**Description:** The adot-operator-images-mirror used [github.com/fsouza/go-dockerclient](https://github.com/fsouza/go-dockerclient), which had a dependency on github.com/containerd/containerd v1.5.5. Switched to the official SDK to remove this dependency.

**Testing:** Ran it against my own AWS account and was able to mirror the images.

### CODE REVIEW

File: tools/release/adot-operator-images-mirror/docker_auth.go [hunk 1]
```diff
@@ -1,22 +1,28 @@
...
+	credentials, err := io.ReadAll(decoder)
 	if err != nil {
-		log.Fatal(err)
+		return "", err
 	}
-
-	creds, ok := authOptions.Configs[registry]
-	if !ok {
-		return nil, fmt.Errorf("no auth found for %s", registry)
+	parts := strings.Split(string(credentials), ":")
```
Comment (sethAmazon) [inline tools/release/adot-operator-images-mirror/docker_auth.go:18]: Any way we can get this as a struct instead of string split by : ?

File: tools/release/adot-operator-images-mirror/main.go [hunk 2]
```diff
@@ -33,20 +33,15 @@ type Repository struct {
...
 // worker is the mirror working function.
-func worker(wg *sync.WaitGroup, workerCh chan []Repository, dc *DockerClient, ecrm *ecrManager) {
+func worker(ctx context.Context, wg *sync.WaitGroup, workerCh chan []Repository, dc *docker.Client, em *ecrManager) {
 	for repos := range workerCh {
 		log.Printf("Starting to mirror repo %s/%s/%s to repo %s/%s", repos[0].Host, repos[0].Registry, repos[0].Name, repos[1].Registry, repos[1].Name)
 
 		m := mirror{
+			ctx:          ctx,
 			dockerClient: dc,
-			ecrManager:   ecrm,
```
Comment (sethAmazon) [inline tools/release/adot-operator-images-mirror/main.go:49]: Why was this renamed?

File: tools/release/adot-operator-images-mirror/docker_auth.go [hunk 3]
```diff
@@ -1,22 +1,28 @@
...
+	authConfig := types.AuthConfig{
+		Username: parts[0],
+		Password: parts[1],
```
Comment (aateeqi) [inline tools/release/adot-operator-images-mirror/docker_auth.go:21]: Do we know for sure that `parts[0]` and `parts[1]` will defined?

File: tools/release/adot-operator-images-mirror/main.go [hunk 4]
```diff
@@ -71,27 +67,25 @@ func main() {
...
+	info, err := dc.Info(ctx)
 	if err != nil {
 		log.Fatalf("Could not get Docker info: %v", err)
 	}
 	log.Printf("Connected to Docker daemon: %s @ %s", info.Name, info.ServerVersion)
 
 	log.Println("Creating AWS client")
-	cfg, err := awsconfig.LoadDefaultConfig(context.TODO())
+	// Override the AWS region with the ecrPublicRegion for ECR authentication.
+	cfg, err := awsconfig.LoadDefaultConfig(ctx, awsconfig.WithRegion(ecrPublicRegion))
```
Comment (sethAmazon) [inline tools/release/adot-operator-images-mirror/main.go:83]: I like that you are add thing region on struct creation instead later in the code

File: tools/release/adot-operator-images-mirror/mirror.go [hunk 5]
```diff
@@ -169,7 +151,7 @@ func (m *mirror) targetRepositoryName() string {
 }
 
 func (m *mirror) getTagResponse(url string) error {
-	req, err := http.NewRequest("GET", url, nil)
+	req, err := http.NewRequest(http.MethodGet, url, nil)
```
Comment (sethAmazon) [inline tools/release/adot-operator-images-mirror/mirror.go:154]: Nice

File: tools/release/adot-operator-images-mirror/mirror.go [hunk 6]
```diff
@@ -178,11 +160,11 @@ func (m *mirror) getTagResponse(url string) error {
 	if err != nil {
 		log.Printf("Failed to get %s, retrying", url)
 		return err
-	} else if res.StatusCode == 429 {
+	} else if res.StatusCode == http.StatusTooManyRequests {
```
Comment (sethAmazon) [inline tools/release/adot-operator-images-mirror/mirror.go:163]: Nice

File: tools/release/adot-operator-images-mirror/mirror.go [hunk 7]
```diff
@@ -178,11 +160,11 @@ func (m *mirror) getTagResponse(url string) error {
...
 		sleepTime := getSleepTime(res.Header.Get("X-RateLimit-Reset"), time.Now())
 		log.Printf("Rate limited on %s, sleeping for %s", url, sleepTime)
 		time.Sleep(sleepTime)
-		return fmt.Errorf("encouter error: too many requests")
+		return fmt.Errorf("encounter error: too many requests")
```
Comment (sethAmazon) [inline tools/release/adot-operator-images-mirror/mirror.go:167]: Nice

**Review by sethAmazon — APPROVED** [review]

File: tools/release/adot-operator-images-mirror/main.go (hunk 2 above)
Comment (jefchien) [reply tools/release/adot-operator-images-mirror/main.go:49]: Just for consistency.

File: tools/release/adot-operator-images-mirror/docker_auth.go (hunk 3 above)
Comment (jefchien) [reply tools/release/adot-operator-images-mirror/docker_auth.go:21]: Yes, based on the [ECR documentation](https://docs.aws.amazon.com/AmazonECR/latest/APIReference/API_AuthorizationData.html):
> When the string is decoded, it is presented in the format user:password for private registry authentication using docker login.

File: tools/release/adot-operator-images-mirror/docker_auth.go (hunk 1 above)
Comment (jefchien) [reply tools/release/adot-operator-images-mirror/docker_auth.go:18]: I don't believe so. That's just the way the authentication token is defined. See https://github.com/aws-observability/aws-otel-collector/pull/771#discussion_r760478940

**Review by aateeqi — APPROVED** [review]

File: tools/release/adot-operator-images-mirror/docker_auth.go (hunk 3 above)
Comment (Aneurysm9) [reply tools/release/adot-operator-images-mirror/docker_auth.go:21]: It is possible that this function was provided a b64-encoded string that is not an auth token and does not contain ':'.  The `strings.Split()` should probably be `strings.Splitn(creds, 2)` to ensure no more than two parts (i.e., if the password contains ':' do not split it into three pieces) and then there should be an assertion that `len(parts) == 2`.

Comment (jefchien) [reply tools/release/adot-operator-images-mirror/docker_auth.go:25]: I've added the check in.
